pkg/cloudprovider/azure.go

```diff
@@ -152,14 +159,59 @@ func (a *Azure) AssignPrivateIP(ip net.IP, node *corev1.Node) error {
 	ipConfigurations := *networkInterface.IPConfigurations
 	name := fmt.Sprintf("%s_%s", node.Name, ipc)
 	untrue := false
+
+	// In some azure setups (Azure private, public ARO, private ARO) outbound connectivity is achieved through
+	// outbound rules tied to the backend address pool of the primary IP of the VM NIC. An Azure constraint
+	// forbids the creation of a secondary IP tied to such address pool and would result in
+	// OutboundRuleCannotBeUsedWithBackendAddressPoolThatIsReferencedBySecondaryIpConfigs.
+	// Work around it by not specifying the backend address pool when an outbound rule is set, even though
+	// that means preventing outbound connectivity to the egress IP, which will be able to reach the
+	// infrastructure subnet nonetheless. In public azure clusters outbound connectivity is achieved through
+	// UserDefinedRouting, which doesn't impose such constraints on secondary IPs.
+	loadBalancerBackendAddressPoolsArgument := (*networkInterface.IPConfigurations)[0].LoadBalancerBackendAddressPools
+	var attachedOutboundRule *network.SubResource
+	if (*networkInterface.IPConfigurations)[0].LoadBalancerBackendAddressPools != nil {
+		for _, ipconfig := range *networkInterface.IPConfigurations {
+			for _, pool := range *ipconfig.LoadBalancerBackendAddressPools {
+				if pool.ID == nil {
+					continue
+				}
+				// for some reason, the struct for the pool above is not entirely filled out:
+				//     BackendAddressPoolPropertiesFormat:(*network.BackendAddressPoolPropertiesFormat)(nil)
+				// Do a separate get for this pool in order to check whether there are any outbound rules
+				// attached to it
+				realPool, err := a.getBackendAddressPool(*pool.ID)
+				if err != nil {
+					return fmt.Errorf("error looking up backend address pool %s with ID %s: %v", *pool.Name, *pool.ID, err)
+				}
+				if realPool.BackendAddressPoolPropertiesFormat != nil {
+					if realPool.BackendAddressPoolPropertiesFormat.OutboundRule != nil {
+						loadBalancerBackendAddressPoolsArgument = nil
+						attachedOutboundRule = realPool.BackendAddressPoolPropertiesFormat.OutboundRule
+						break
```

Comment:
This will break out of the inner for loop (line 175). Should it also break us out of the outer for loop (line 174)? If so, maybe we could add a check for `loadBalancerBackendAddressPoolsArgument` in line 199?

or use a label? Example:
```
NeedleSearch:
	for _, stack := range haystack {
		for _, straw := range stack {
			if straw == 10 { // Found the needle!
				fmt.Println("Found the needle!")
				break NeedleSearch
			}
		}
	}
```